### ISSUE TITLE
Unable to parse MDT strings that contain cyrillic

### ISSUE THREAD

**Issue by Wotuu** [Zireael-N/cli-weakauras-parser#1]
Hey,

I've gotten some reports of users where parsing MDT strings that contain certain cyrillic characters in their titles or comments fail to parse. I'm using the MDT Lua based implementation on keystone.guru which has this issue. I just converted a host of things to implement this weakauras parser since I still had it on my TODO list but it appears to suffer from the same or a similar issue.

Take this string for example:

```
!DF1YUroru0)LzfibwUE4QCTLbwasygPM1w5rBh6bpDJsChcSikpemcrGzfsSbbly2NKjzMm55Vq5)iU36HDz3EblqZcuu6U6Y19EpNZ9uv1Doj)lZN8SPZ(K5RVrvX08nYNS5IQfB)K1RkQRl(05LlWvS(Y6fpgNF28TY)ObRz6E5z8868j1f7bV2CyZr67BoqFV(o9P7VgoQ54MJGrhPVeFA(KPZklNT5YQ6VlpJKKp5Blk(68S08jl24PfBwVduZmd0MAhLX48igjoLPuPCvsEgnp7d584ioLrvK4KujUkZA5acZGKQ)9Md1paW4MMt0VP5y9L6BbqCmaO7AoE)1GXphHQ(my0H6Z0x1CuZj7VM(A9PimHr)L(uiYtAEbeZLaDUdw9TyS6lG88dnhhbd)ni4B0xHPg(05yW47W)VQ5aOSpa)Dj8jd3nf4k8jqQUglm85)auONdG6NraIQM5Px08JiKrLDoqPAK0bscpHgrzGGWvKuQxsKmysASKisejSvKK)eQeI2lr0Iuw)wt)6CtN6xmYXR03dO4bG6xJZcp7E9B0Ndy60MFcM9TU(OrQG)FPDLixdWkRhwtHoflrjs4sUIzXkLlJsekjWbMKAXk7FB77DRWZdjtsmlItOIyfvk4elzyCreJhRIzcIY5fPHKbY(RH6FhGMdGADluNxdJrx29GGEnqZFfWIZd(iycC)Z5aqGaEKdVGmdC6MMxaXFva(s6HpEAeHelKmoi4uh(IJJyPugNiPkM6))7ve98FmrevQILGfmfzVX)rsJ4jcMiwgRsPoZ3)zsY7w6k75aOSijCQixOuaL90LebhjeZttaPiDf6(3yvb6DiGHBDo0Z0N(bMYIG8cObB1Gd9Cewcska6rDybgS76vllqeT5YT3UyE9twwvbfmUDIjlHRB2TOcH7KTNvw)fLL7u0DWV7Wo35iUDGoJUR56iDnEVH5nL9nsS7DxEi2ev3w7pE58Tkwm3CVfvHxBv8m8En4UVDkQGRFMTyU)ANymUVbqFlY8VKz34tDhB5a8OpG1(GA3vN5tklzfLLbHzxIWDQbgBsys5EriScobzyUXmRk8voSIScR25JOlN25KHvHBQIO7CzdI6kIYrade8C0Z8qMwUbRW31cXiIgsiCYiYvzijPpGLd4Kkml2isxjloxsy1bW3LBAVQZg0fCMRUeZBzzxPz9aBFfq5uaJ3mJOgTP5negJBMCK1SoXiAEtDw6arV9qDRDXvnBoYgRODUqIPbQ6LqERfEvNfHhcs6GC6azsiitcStPMKl7y7GC0zUCYlzuW7nwUghzeT35Z8UIvmo0EmM0YymaRAWdfrYy7LDUjC3ACVAyA7zHEtCngy2Z45bHD3FIbeYr9g0ENwKMhSDvoQrHAamnEuA26uOUJxP2ZmhypSjXx962aCVgSHPLEbMYvCpom5Lu6WwGTajTfOdxDRnSGI(XZALVUcnU)0gzBns7XDNfSl9IaPl1DCOTx1bqCKSNYyV8vmsbC23obt0dYSE817mQBtKRqTRS1MsBLVv2f0rgI9emI9ogV7WNFJXVDwrx)G4fw3LZu3fXodYGDxWdyX9sL0u108v270pt4s594FxGysn)0Wb7YOHNVHlYekl8iWv2oz(wbWV(ftx52ffF1IQPF2IzZT)qy)myWt2OyRLZQWVMGnvlNnnFsYUIV)PV3Jv795Vpmx()8d
```

This results in the following output (I write the above string to a ram disk and point the cli_weakauras_parser to that since I couldn't get it to work in any other way):

```
The command \"'cli_weakauras_parser' 'decode' '/dev/shm/keystone.guru/mdt/1067122055'\" failed.
Exit Code: 1(General error)
Working directory: /home/vagrant/Git/private/keystone.guru/public
Output:
================

Error Output:
================
Error: Invalid base64 length
```

The full string contents are actually the following if you want to play around with it.
```json
{"objects":{"1":{"n":true,"d":{"1":344.31083998495,"2":-440.42329105871,"3":1,"4":true,"5":"\u0418\u0441\u043f\u043e\u043b\u044c\u0437\u0443\u0435\u043c \u0443\u0440\u043d\u0443  \u0447\u0442\u043e\u0431 \u0441\u0431\u0438\u0442\u044c \u043a\u0430\u0441\u0442 \u0420\u0430\u0437\u044c\u044f\u0440\u0435\u043d\u043d\u043e\u043c\u0443 \u0434\u0443\u0445\u0443. \u0415\u0441\u043b\u0438 \u0443 \u0432\u0430\u0441 \u0432 \u0433\u0440\u0443\u043f\u043f\u0435 \u0435\u0441\u0442\u044c \u0438\u0433\u0440\u043e\u043a \u0441 \u041d\u043e\u0447\u043d\u044b\u043c \u043d\u0430\u0440\u043e\u0434\u0446\u0435\u043c"}},"2":{"n":true,"d":{"1":452.23399491825,"2":-473.23207165653,"3":1,"4":true,"5":"\u041f\u043e\u0441\u043b\u0435 \u0443\u0431\u0438\u0439\u0441\u0442\u0432\u0430 \u044d\u0442\u043e\u0433\u043e \u043f\u0430\u043a\u0430 \u0432\u043e\u0437\u0432\u0440\u0430\u0449\u0430\u0439\u0442\u0435\u0441\u044c \u043a \u0425\u0430\u043a\u0430\u0440\u0443"}},"3":{"n":true,"d":{"1":480.43596547493,"2":-247.56973393723,"3":3,"4":true,"5":"\u0418\u0441\u043f\u043e\u043b\u044c\u0437\u0443\u0435\u043c \u0443\u0440\u043d\u0443. \u0415\u0441\u043b\u0438 \u0443 \u0432\u0430\u0441 \u0432 \u0433\u0440\u0443\u043f\u043f\u0435 \u0435\u0441\u0442\u044c \u0438\u0433\u0440\u043e\u043a \u0441 \u041d\u043e\u0447\u043d\u044b\u043c \u043d\u0430\u0440\u043e\u0434\u0446\u0435\u043c"}},"4":{"n":true,"d":{"1":503.41260927641,"2":-346.34090361971,"3":2,"4":true,"5":"\u0418\u043d\u0436\u0438\u043d\u0435\u0440 \u043c\u043e\u0436\u0435\u0442 \u043e\u0442\u043a\u043b\u044e\u0447\u0438\u0442\u044c \"\u043a\u043e\u043d\u0432\u0435\u0435\u0440\" \u0441 \u043a\u0430\u043f\u043b\u044f\u043c\u0438"}},"5":{"n":true,"d":{"1":548.11067349652,"2":-300.38234172939,"3":1,"4":true,"5":"\u0418\u0441\u043f\u043e\u043b\u044c\u0437\u0443\u0435\u043c \u0443\u0440\u043d\u0443  \u0447\u0442\u043e\u0431 \u0441\u0431\u0438\u0442\u044c \u043a\u0430\u0441\u0442 \u0420\u0430\u0437\u044c\u044f\u0440\u0435\u043d\u043d\u043e\u043c\u0443 \u0434\u0443\u0445\u0443. \u0415\u0441\u043b\u0438 \u0443 \u0432\u0430\u0441 \u0432 \u0433\u0440\u0443\u043f\u043f\u0435 \u0435\u0441\u0442\u044c \u0438\u0433\u0440\u043e\u043a \u0441 \u041d\u043e\u0447\u043d\u044b\u043c \u043d\u0430\u0440\u043e\u0434\u0446\u0435\u043c"}},"6":{"n":true,"d":{"1":436.27907596839,"2":-218.45636070982,"3":4,"4":true,"5":"\u0418\u0441\u043f\u043e\u043b\u044c\u0437\u0443\u0435\u043c \u0443\u0440\u043d\u0443  \u0447\u0442\u043e\u0431 \u0441\u0431\u0438\u0442\u044c \u043a\u0430\u0441\u0442. \u0415\u0441\u043b\u0438 \u0443 \u0432\u0430\u0441 \u0432 \u0433\u0440\u0443\u043f\u043f\u0435 \u0435\u0441\u0442\u044c \u0438\u0433\u0440\u043e\u043a \u0441 \u041d\u043e\u0447\u043d\u044b\u043c \u043d\u0430\u0440\u043e\u0434\u0446\u0435\u043c"}},"7":{"n":true,"d":{"1":523.79104699759,"2":-211.65304852798,"3":4,"4":true,"5":"\u0424\u043e\u043a\u0443\u0441\u0438\u043c \u043c\u043e\u0431\u0430, \u043d\u0435 \u0434\u0430\u0435\u043c \u0441\u043a\u0430\u0441\u0442\u043e\u0432\u0430\u0442\u044c."}}},"text":"\u0441\u0442\u043e\u0440\u043e\u043d\u0430 \u0440\u043e\u0443\u0442 \u0442\u0435\u0441\u0442","week":8,"mdi":{"freeholdJoined":false,"freehold":1,"beguiling":1},"difficulty":15,"value":{"riftOffsets":{"1":[],"2":[],"3":[],"4":[],"5":[],"6":[],"7":[],"8":[],"9":[],"10":[],"11":[],"12":[]},"currentPull":30,"selection":{"1":30},"pulls":{"1":{"1":{"1":1,"2":2,"3":3},"2":{"1":1,"2":2,"3":3},"3":{"1":1},"color":"ff3eff"},"2":{"1":{"1":6,"2":4,"3":5},"2":{"1":4},"4":{"1":1,"2":2},"color":"3eff9e","5":{"1":1}},"3":{"color":"ff3e3e","6":{"1":1}},"4":{"1":{"1":7},"4":{"1":4,"2":6,"3":3,"4":5},"color":"3e9eff"},"5":{"color":"fffb3e","3":{"1":2,"2":3}},"6":{"16":{"1":1},"17":{"1":1,"2":2},"color":"3eff3e","15":{"1":1}},"7":{"16":{"1":2},"color":"ff3e9e","18":{"1":1,"2":2}},"8":{"17":{"1":3,"2":4,"3":5},"color":"3effff","15":{"1":2}},"9":{"16":{"1":3},"color":"ff9b3e","18":{"1":4,"2":3},"15":{"1":3}},"10":{"color":"3e3eff","19":{"1":1}},"11":{"color":"a1ff3e","7":{"1":1}},"12":{"8":{"1":2,"2":3,"3":1,"4":4},"9":{"1":1},"10":{"1":1},"color":"ff3eff"},"13":{"color":"3eff9e","9":{"1":2,"2":4,"3":3}},"14":{"color":"ff3e3e","7":{"1":2}},"15":{"8":{"1":5,"2":6,"3":8,"4":7},"color":"3e9eff","11":{"1":2}},"16":{"color":"fffb3e","11":{"1":1}},"17":{"color":"3eff3e","11":{"1":3}},"18":{"color":"ff3e9e","12":{"1":2,"2":1,"3":3}},"19":{"13":{"1":1},"14":{"1":1},"color":"3effff"},"20":{"color":"ff9b3e","6":{"1":2}},"21":{"5":{"1":2},"2":{"1":6,"2":5,"3":7},"color":"3e3eff"},"22":{"1":{"1":8},"4":{"1":7},"color":"a1ff3e"},"23":{"24":[],"color":"ff3eff","20":{"1":2,"2":1,"3":3,"4":4},"21":[],"22":[],"23":[]},"24":{"24":{"1":3,"2":4},"color":"3eff9e","21":{"1":2,"2":3},"22":{"1":2,"2":1},"23":{"1":1}},"25":{"24":[],"26":{"1":2,"2":13,"3":7},"color":"ff3e3e","21":[],"22":{"1":3,"2":4},"23":{"1":2}},"26":{"24":[],"color":"3e9eff","28":[],"21":[],"26":{"1":6,"2":1,"3":8,"4":5,"5":3}},"27":{"24":{"1":9,"2":6},"color":"fffb3e","28":[],"21":{"1":6},"22":{"1":6,"2":5},"23":{"1":3}},"28":{"24":{"1":12},"25":[],"color":"3eff3e","27":{"1":3},"20":{"1":5},"26":{"1":14,"2":15}},"29":{"25":{"1":7},"26":[],"20":{"1":6},"color":"ff3e9e","22":[],"23":[]},"30":{"color":"3effff","20":{"1":7,"2":8},"26":[],"22":{"1":7,"2":8},"23":{"1":4}},"31":{"color":"ff9b3e","29":{"1":1}},"32":{"30":{"1":1},"color":"3e3eff"}},"teeming":false,"currentSublevel":1,"currentDungeonIdx":29},"colorPaletteInfo":{"autoColoring":true,"colorPaletteIdx":4},"mdiEnabled":false,"uid":"5v6zj(C9xM)"}
```

This string actually parses correctly on Keystone.guru at the moment (try to import it). If you import it and try to share it again the string generation will fail (and I strip it of cyrillic characters and try again because of the previously mentioned bug).

Either way - I hope that's enough info for you, let me know if I can assist in any other way.

**Comment by Zireael-N** [comment]
Hey!

I just tried to decode your string using the following command:

```bash
echo -n '!DF1YUroru0)LzfibwUE4QCTLbwasygPM1w5rBh6bpDJsChcSikpemcrGzfsSbbly2NKjzMm55Vq5)iU36HDz3EblqZcuu6U6Y19EpNZ9uv1Doj)lZN8SPZ(K5RVrvX08nYNS5IQfB)K1RkQRl(05LlWvS(Y6fpgNF28TY)ObRz6E5z8868j1f7bV2CyZr67BoqFV(o9P7VgoQ54MJGrhPVeFA(KPZklNT5YQ6VlpJKKp5Blk(68S08jl24PfBwVduZmd0MAhLX48igjoLPuPCvsEgnp7d584ioLrvK4KujUkZA5acZGKQ)9Md1paW4MMt0VP5y9L6BbqCmaO7AoE)1GXphHQ(my0H6Z0x1CuZj7VM(A9PimHr)L(uiYtAEbeZLaDUdw9TyS6lG88dnhhbd)ni4B0xHPg(05yW47W)VQ5aOSpa)Dj8jd3nf4k8jqQUglm85)auONdG6NraIQM5Px08JiKrLDoqPAK0bscpHgrzGGWvKuQxsKmysASKisejSvKK)eQeI2lr0Iuw)wt)6CtN6xmYXR03dO4bG6xJZcp7E9B0Ndy60MFcM9TU(OrQG)FPDLixdWkRhwtHoflrjs4sUIzXkLlJsekjWbMKAXk7FB77DRWZdjtsmlItOIyfvk4elzyCreJhRIzcIY5fPHKbY(RH6FhGMdGADluNxdJrx29GGEnqZFfWIZd(iycC)Z5aqGaEKdVGmdC6MMxaXFva(s6HpEAeHelKmoi4uh(IJJyPugNiPkM6))7ve98FmrevQILGfmfzVX)rsJ4jcMiwgRsPoZ3)zsY7w6k75aOSijCQixOuaL90LebhjeZttaPiDf6(3yvb6DiGHBDo0Z0N(bMYIG8cObB1Gd9Cewcska6rDybgS76vllqeT5YT3UyE9twwvbfmUDIjlHRB2TOcH7KTNvw)fLL7u0DWV7Wo35iUDGoJUR56iDnEVH5nL9nsS7DxEi2ev3w7pE58Tkwm3CVfvHxBv8m8En4UVDkQGRFMTyU)ANymUVbqFlY8VKz34tDhB5a8OpG1(GA3vN5tklzfLLbHzxIWDQbgBsys5EriScobzyUXmRk8voSIScR25JOlN25KHvHBQIO7CzdI6kIYrade8C0Z8qMwUbRW31cXiIgsiCYiYvzijPpGLd4Kkml2isxjloxsy1bW3LBAVQZg0fCMRUeZBzzxPz9aBFfq5uaJ3mJOgTP5negJBMCK1SoXiAEtDw6arV9qDRDXvnBoYgRODUqIPbQ6LqERfEvNfHhcs6GC6azsiitcStPMKl7y7GC0zUCYlzuW7nwUghzeT35Z8UIvmo0EmM0YymaRAWdfrYy7LDUjC3ACVAyA7zHEtCngy2Z45bHD3FIbeYr9g0ENwKMhSDvoQrHAamnEuA26uOUJxP2ZmhypSjXx962aCVgSHPLEbMYvCpom5Lu6WwGTajTfOdxDRnSGI(XZALVUcnU)0gzBns7XDNfSl9IaPl1DCOTx1bqCKSNYyV8vmsbC23obt0dYSE817mQBtKRqTRS1MsBLVv2f0rgI9emI9ogV7WNFJXVDwrx)G4fw3LZu3fXodYGDxWdyX9sL0u108v270pt4s594FxGysn)0Wb7YOHNVHlYekl8iWv2oz(wbWV(ftx52ffF1IQPF2IzZT)qy)myWt2OyRLZQWVMGnvlNnnFsYUIV)PV3Jv795Vpmx()8d' | ./cli_weakauras_parser decode -
```

and it worked correctly.

Is it possible that your input is terminated with `\r\n` or `\n`? This tool does not strip whitespace on its own.

**Comment by Zireael-N** [comment]
Just noticed that you are writing MDT strings to a RAM disk. If you could share problematic files, that would be helpful.

**Comment by Wotuu** [comment]
Thanks for your swift reply! Turns out I made an oopsie when writing the file to disk - I was always encoding the string to JSON since I'm re-using a bit of code. But that step shouldn't be there for the encoding stage, so it was wrapping the string in quotes when it wrote it to disk. The library would then of course fail since the quotes aren't base64 spec. Whoops.

I've got it working now, thanks so much for your help!
